# Copyright (c) 2023, Auterion AG
# SPDX-License-Identifier: BSD-3-Clause

diff --git a/platforms/nuttx/NuttX/nuttx/drivers/Makefile b/platforms/nuttx/NuttX/nuttx/drivers/Makefile
index e2e0218490..57476761d7 100644
--- a/platforms/nuttx/NuttX/nuttx/drivers/Makefile
+++ b/platforms/nuttx/NuttX/nuttx/drivers/Makefile
@@ -73,6 +73,7 @@ include segger/Make.defs
 ifeq ($(CONFIG_SPECIFIC_DRIVERS),y)
 -include platform/Make.defs
 endif
+include itm/Make.defs

 AOBJS = $(ASRCS:.S=$(OBJEXT))
 COBJS = $(CSRCS:.c=$(OBJEXT))
