# Copyright (c) 2023, Auterion AG
# SPDX-License-Identifier: BSD-3-Clause

diff --git a/platforms/nuttx/NuttX/nuttx/sched/semaphore/sem_post.c b/platforms/nuttx/NuttX/nuttx/sched/semaphore/sem_post.c
index e9f69dfe1a..af427937e4 100644
--- a/platforms/nuttx/NuttX/nuttx/sched/semaphore/sem_post.c
+++ b/platforms/nuttx/NuttX/nuttx/sched/semaphore/sem_post.c
@@ -33,6 +33,7 @@

 #include "sched/sched.h"
 #include "semaphore/semaphore.h"
+#include <nuttx/itm/itm.h>

 /****************************************************************************
  * Public Functions
@@ -69,6 +70,7 @@

 int nxsem_post(FAR sem_t *sem)
 {
+  EMDBG_LOG_SEMAPHORE_POST(sem);
   FAR struct tcb_s *stcb = NULL;
   irqstate_t flags;
   int ret = -EINVAL;
diff --git a/platforms/nuttx/NuttX/nuttx/sched/semaphore/sem_wait.c b/platforms/nuttx/NuttX/nuttx/sched/semaphore/sem_wait.c
index eec67723d7..fea433e2f5 100644
--- a/platforms/nuttx/NuttX/nuttx/sched/semaphore/sem_wait.c
+++ b/platforms/nuttx/NuttX/nuttx/sched/semaphore/sem_wait.c
@@ -34,6 +34,7 @@

 #include "sched/sched.h"
 #include "semaphore/semaphore.h"
+#include <nuttx/itm/itm.h>

 /****************************************************************************
  * Public Functions
@@ -69,6 +70,7 @@

 int nxsem_wait(FAR sem_t *sem)
 {
+  EMDBG_LOG_SEMAPHORE_WAIT(sem);
   FAR struct tcb_s *rtcb = this_task();
   irqstate_t flags;
   int ret = -EINVAL;
diff --git a/platforms/nuttx/NuttX/nuttx//sched/sched/sched_addreadytorun.c b/platforms/nuttx/NuttX/nuttx//sched/sched/sched_addreadytorun.c
index 1b2abd47a3..7551b92366 100644
--- a/platforms/nuttx/NuttX/nuttx//sched/sched/sched_addreadytorun.c
+++ b/platforms/nuttx/NuttX/nuttx//sched/sched/sched_addreadytorun.c
@@ -27,6 +27,7 @@
 #include <stdbool.h>
 #include <queue.h>
 #include <assert.h>
+#include <nuttx/itm/itm.h>

 #include "irq/irq.h"
 #include "sched/sched.h"
@@ -108,6 +109,7 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
       ret = false;
     }

+  if (!ret) emdbg_itm16_block(EMDBG_TASK_RUNNABLE, btcb->pid);
   return ret;
 }
 #endif /* !CONFIG_SMP */
diff --git a/platforms/nuttx/NuttX/nuttx/sched/task/task_prctl.c b/platforms/nuttx/NuttX/nuttx/sched/task/task_prctl.c
index 017ee5b55b..bcd193ecc0 100644
--- a/platforms/nuttx/NuttX/nuttx/sched/task/task_prctl.c
+++ b/platforms/nuttx/NuttX/nuttx/sched/task/task_prctl.c
@@ -31,6 +31,7 @@
 #include <debug.h>

 #include <nuttx/sched.h>
+#include <nuttx/itm/itm.h>

 #include "sched/sched.h"
 #include "task/task.h"
@@ -128,6 +129,7 @@ int prctl(int option, ...)

               strncpy(tcb->name, name, CONFIG_TASK_NAME_SIZE);
               tcb->name[CONFIG_TASK_NAME_SIZE] = '\0';
+              EMDBG_LOG_TASK_START(tcb);
             }
           else
             {
diff --git a/platforms/nuttx/src/px4/common/cpuload.cpp b/platforms/nuttx/src/px4/common/cpuload.cpp
index 62c5c6205d..2c22732b32 100644
--- a/platforms/nuttx/src/px4/common/cpuload.cpp
+++ b/platforms/nuttx/src/px4/common/cpuload.cpp
@@ -48,6 +48,7 @@
 #if defined(__PX4_NUTTX) && defined(CONFIG_SCHED_INSTRUMENTATION)
 __BEGIN_DECLS
 # include <nuttx/sched_note.h>
+# include <nuttx/itm/itm.h>
 
 __EXPORT struct system_load_s system_load;
 
@@ -110,6 +111,7 @@ void cpuload_initialize_once()
 
 void sched_note_start(FAR struct tcb_s *tcb)
 {
+  EMDBG_LOG_TASK_START(tcb);
   // find first free slot
   if (system_load.initialized) {
      for (auto &task : system_load.tasks) {
@@ -128,6 +130,7 @@ void sched_note_start(FAR struct tcb_s *tcb)
 
 void sched_note_stop(FAR struct tcb_s *tcb)
 {
+  emdbg_itm16_block(EMDBG_TASK_STOP, tcb->pid);
   if (system_load.initialized) {
      for (auto &task : system_load.tasks) {
         if (task.tcb && task.tcb->pid == tcb->pid) {
@@ -143,8 +146,11 @@ void sched_note_stop(FAR struct tcb_s *tcb)
   }
 }
 
+static uint8_t emdbg_prev_state = 0;
 void sched_note_suspend(FAR struct tcb_s *tcb)
 {
+  emdbg_prev_state = tcb->task_state;
+  emdbg_itm16_block(EMDBG_TASK_SUSPEND, tcb->pid);
   if (system_load.initialized) {
      if (tcb->pid == 0) {
         system_load.tasks[0].total_runtime += hrt_elapsed_time(&system_load.tasks[0].curr_start_time);
@@ -170,6 +176,7 @@ void sched_note_suspend(FAR struct tcb_s *tcb)
 
 void sched_note_resume(FAR struct tcb_s *tcb)
 {
+  EMDBG_LOG_TASK_RESUME(tcb, emdbg_prev_state);
   if (system_load.initialized) {
      if (tcb->pid == 0) {
         hrt_store_absolute_time(&system_load.tasks[0].curr_start_time);
@@ -191,5 +198,49 @@ void sched_note_resume(FAR struct tcb_s *tcb)
      }
   }
 }
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_PREEMPTION
+void sched_note_premption(FAR struct tcb_s *tcb, bool locked)
+{
+  const uint8_t channel = locked ? EMDBG_PREEMPTION_LOCK : EMDBG_PREEMPTION_UNLOCK;
+  emdbg_itm8(channel, tcb->lockcount);
+}
+#endif
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_CSECTION
+void sched_note_csection(FAR struct tcb_s *tcb, bool enter)
+{
+  const uint8_t channel = enter ? EMDBG_CSECTION_ENTER : EMDBG_CSECTION_LEAVE;
+  emdbg_itm8(channel, tcb->irqcount);
+}
+#endif
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SPINLOCKS
+static inline void note_spincommon(FAR volatile void *spinlock, int type)
+{
+  emdbg_itm16(type, (uint32_t)spinlock >> 3);
+}
+
+void sched_note_spinlock(FAR struct tcb_s *tcb, FAR volatile void *spinlock)
+{
+  note_spincommon(spinlock, EMDBG_SPINLOCK_LOCK);
+}
+
+void sched_note_spinlocked(FAR struct tcb_s *tcb, FAR volatile void *spinlock)
+{
+  note_spincommon(spinlock, EMDBG_SPINLOCK_LOCKED);
+}
+
+void sched_note_spinunlock(FAR struct tcb_s *tcb, FAR volatile void *spinlock)
+{
+  note_spincommon(spinlock, EMDBG_SPINLOCK_UNLOCK);
+}
+
+void sched_note_spinabort(FAR struct tcb_s *tcb, FAR volatile void *spinlock)
+{
+  note_spincommon(spinlock, EMDBG_SPINLOCK_ABORT);
+}
+#endif
+
 __END_DECLS
 #endif // PX4_NUTTX && CONFIG_SCHED_INSTRUMENTATION

diff --git a/platforms/common/px4_work_queue/WorkQueue.cpp b/platforms/common/px4_work_queue/WorkQueue.cpp
index fa32017912..1b2aa522b1 100644
--- a/platforms/common/px4_work_queue/WorkQueue.cpp
+++ b/platforms/common/px4_work_queue/WorkQueue.cpp
@@ -39,6 +39,7 @@
 #include <px4_platform_common/tasks.h>
 #include <px4_platform_common/time.h>
 #include <drivers/drv_hrt.h>
+#include <nuttx/itm/itm.h>

 namespace px4
 {
@@ -183,8 +184,10 @@ void WorkQueue::Run()
                        WorkItem *work = _q.pop();

                        work_unlock(); // unlock work queue to run (item may requeue itself)
+                       EMDBG_WORKQUEUE_START(work);
                        work->RunPreamble();
                        work->Run();
+                       EMDBG_WORKQUEUE_STOP(work);
                        // Note: after Run() we cannot access work anymore, as it might have been deleted
                        work_lock(); // re-lock
                }
diff --git a/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_free.c b/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_free.c
index 5513ba1772..8cf297e747 100644
--- a/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_free.c
+++ b/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_free.c
@@ -29,6 +29,7 @@
 
 #include <nuttx/arch.h>
 #include <nuttx/mm/mm.h>
+#include <nuttx/itm/itm.h>
 
 #include "mm_heap/mm.h"
 #include "kasan/kasan.h"
@@ -94,6 +95,7 @@ void mm_free(FAR struct mm_heap_s *heap, FAR void *mem)
       mm_add_delaylist(heap, mem);
       return;
     }
+  EMDBG_HEAP_FREE(mem);
 
   kasan_poison(mem, mm_malloc_size(mem));
 
diff --git a/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_initialize.c b/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_initialize.c
index 3b5ac7ed8a..f65319bafb 100644
--- a/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_initialize.c
+++ b/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_initialize.c
@@ -29,6 +29,7 @@
 #include <debug.h>
 
 #include <nuttx/mm/mm.h>
+#include <nuttx/itm/itm.h>
 
 #include "mm_heap/mm.h"
 #include "kasan/kasan.h"
@@ -152,6 +153,7 @@ void mm_addregion(FAR struct mm_heap_s *heap, FAR void *heapstart,
   mm_addfreechunk(heap, node);
 
   mm_givesemaphore(heap);
+  EMDBG_HEAP_ADDREGION(heapbase, heapend - heapbase);
 }
 
 /****************************************************************************
diff --git a/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_malloc.c b/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_malloc.c
index a38633bbfe..d7594d860c 100644
--- a/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_malloc.c
+++ b/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_malloc.c
@@ -32,6 +32,7 @@
 #include <nuttx/arch.h>
 #include <nuttx/mm/mm.h>
 #include <nuttx/sched.h>
+#include <nuttx/itm/itm.h>
 
 #include "mm_heap/mm.h"
 #include "kasan/kasan.h"
@@ -117,6 +118,7 @@ FAR void *mm_malloc(FAR struct mm_heap_s *heap, size_t size)
 
   if (size < 1)
     {
+      EMDBG_HEAP_MALLOC_RESULT(0);
       return NULL;
     }
 
@@ -125,10 +127,12 @@ FAR void *mm_malloc(FAR struct mm_heap_s *heap, size_t size)
    */
 
   alignsize = MM_ALIGN_UP(size + SIZEOF_MM_ALLOCNODE);
+  EMDBG_HEAP_MALLOC(size);
   if (alignsize < size)
     {
       /* There must have been an integer overflow */
 
+      EMDBG_HEAP_MALLOC_RESULT(0);
       return NULL;
     }
 
@@ -234,6 +238,7 @@ FAR void *mm_malloc(FAR struct mm_heap_s *heap, size_t size)
 
   DEBUGASSERT(ret == NULL || mm_heapmember(heap, ret));
   mm_givesemaphore(heap);
+  EMDBG_HEAP_MALLOC_RESULT(ret);
 
   if (ret)
     {
diff --git a/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_memalign.c b/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_memalign.c
index 47c27d61ba..ad4cad92f0 100644
--- a/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_memalign.c
+++ b/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_memalign.c
@@ -27,6 +27,7 @@
 #include <assert.h>
 
 #include <nuttx/mm/mm.h>
+#include <nuttx/itm/itm.h>
 
 #include "mm_heap/mm.h"
 #include "kasan/kasan.h"
@@ -228,5 +229,6 @@ FAR void *mm_memalign(FAR struct mm_heap_s *heap, size_t alignment,
   kasan_unpoison((FAR void *)alignedchunk,
                  mm_malloc_size((FAR void *)alignedchunk));
 
+  EMDBG_HEAP_MEMALIGN(rawchunk, size, alignedchunk);
   return (FAR void *)alignedchunk;
 }
diff --git a/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_realloc.c b/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_realloc.c
index 6331465d34..c225edd2ec 100644
--- a/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_realloc.c
+++ b/platforms/nuttx/NuttX/nuttx/mm/mm_heap/mm_realloc.c
@@ -30,6 +30,7 @@
 #include <assert.h>
 
 #include <nuttx/mm/mm.h>
+#include <nuttx/itm/itm.h>
 
 #include "mm_heap/mm.h"
 #include "kasan/kasan.h"
@@ -134,6 +135,7 @@ FAR void *mm_realloc(FAR struct mm_heap_s *heap, FAR void *oldmem,
 
       MM_ADD_BACKTRACE(heap, oldnode);
 
+      EMDBG_HEAP_REALLOC(oldmem, size, oldmem);
       return oldmem;
     }
 
@@ -349,6 +351,7 @@ FAR void *mm_realloc(FAR struct mm_heap_s *heap, FAR void *oldmem,
           memcpy(newmem, oldmem, oldsize - SIZEOF_MM_ALLOCNODE);
         }
 
+      EMDBG_HEAP_REALLOC(oldmem, size, newmem);
       return newmem;
     }
 
